Cython/Utility/ObjectHandling.c

```diff
@@ -234,15 +234,15 @@ static CYTHON_INLINE PyObject *__Pyx_PyIter_Next2(PyObject* iterator, PyObject*
     // We always do a quick slot check because calling PyIter_Check() is so wasteful.
     iternextfunc iternext = __Pyx_PyObject_TryGetSlot(iterator, tp_iternext, iternextfunc);
     if (likely(iternext)) {
-#if CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY
-        next = iternext(iterator);
-        if (likely(next))
-            return next;
+#if CYTHON_USE_TYPE_SLOTS
 #if CYTHON_COMPILING_IN_CPYTHON && PY_VERSION_HEX < 0x030d0000
         if (unlikely(iternext == &_PyObject_NextNotImplemented))
             return NULL;
 #endif
```

Comment:
This is only slightly related but spotted at the same time - I think it only makes sense to do this check before trying to call `iternext`

Reply:
I think I deliberately did it afterwards because it's actually unlikely to find this function in `tp_iternext`. I'd expect the slot to be either `NULL` or useful, not `_PyObject_NextNotImplemented`. But if calling `tp_iternext` fails (with an exception), then checking for `_PyObject_NextNotImplemented` is a reasonable thing to do here because we know that we can stop trying.

Note that your change is actually invalid because it now returns `NULL` without setting an exception.

Reply:
Makes sense. I've reverted that